File: dotnet/DesignPatterns/BehavioralDesignPatterns/MediatorDesignPattern/MediatorDesignPattern.Example3/Auctioneer.cs
namespace MediatorDesignPattern.Example3;

/// <summary>
/// Concrete Mediator - Auctioneer that coordinates all bidding activities
/// </summary>
public class Auctioneer : IAuctionMediator
{
    private readonly List<Bidder> _bidders = new();
    private readonly string _itemName;
    private decimal _currentHighestBid;
    private Bidder? _highestBidder;

    public Auctioneer(string itemName, decimal startingPrice)
    {
        _itemName = itemName;
        _currentHighestBid = startingPrice;
        Console.WriteLine($"[AUCTIONEER] Auction started for: {_itemName}");
        Console.WriteLine($"[AUCTIONEER] Starting price: ${startingPrice}");
    }

    public void RegisterBidder(Bidder bidder)
    {
        _bidders.Add(bidder);
        Console.WriteLine($"[AUCTIONEER] {bidder.Name} joined the auction.");
    }

    public void PlaceBid(Bidder bidder, decimal amount)
    {
        if (amount <= _currentHighestBid)
        {
            bidder.ReceiveNotification($"Bid rejected. Must be higher than ${_currentHighestBid}");
            return;
        }

        _currentHighestBid = amount;
        _highestBidder = bidder;

        // Notify all other bidders about the new highest bid
        foreach (var b in _bidders.Where(b => b != bidder))
        {
            b.ReceiveNotification($"New highest bid: ${amount} by {bidder.Name}");
        }

        Console.WriteLine($"[AUCTIONEER] New highest bid: ${amount} by {bidder.Name}");
    }

    public void AnnounceWinner()
    {
        Console.WriteLine("\n[AUCTIONEER] Auction ended!");
        
        if (_highestBidder == null)
        {
            Console.WriteLine("[AUCTIONEER] No bids were placed. Item not sold.");
            return;
        }

        Console.WriteLine($"[AUCTIONEER] Winner: {_highestBidder.Name}");
        Console.WriteLine($"[AUCTIONEER] Winning bid: ${_currentHighestBid}");
        Console.WriteLine($"[AUCTIONEER] Item: {_itemName}");

        // Notify all bidders about the result
        foreach (var bidder in _bidders)
        {
            var message = bidder == _highestBidder 
                ? $"Congratulations! You won {_itemName} for ${_currentHighestBid}" 
                : $"{_itemName} sold to {_highestBidder.Name} for ${_currentHighestBid}";
            bidder.ReceiveNotification(message);
        }
    }
}

File: nodejs/DesignPatterns/BehavioralDesignPatterns/MediatorDesignPattern/MediatorDesignPattern.Example3/src/Auctioneer.ts
import { IAuctionMediator } from './IAuctionMediator';
import { Bidder } from './Bidder';

export class Auctioneer implements IAuctionMediator {
  private readonly _bidders: Bidder[] = [];
  private readonly _itemName: string;
  private _currentHighestBid: number;
  private _highestBidder: Bidder | undefined;

  constructor(itemName: string, startingPrice: number) {
    this._itemName = itemName;
    this._currentHighestBid = startingPrice;
    console.log(`[AUCTIONEER] Auction started for: ${this._itemName}`);
    console.log(`[AUCTIONEER] Starting price: $${startingPrice}`);
  }

  registerBidder(bidder: Bidder): void {
    this._bidders.push(bidder);
    console.log(`[AUCTIONEER] ${bidder.name} joined the auction.`);
  }

  placeBid(bidder: Bidder, amount: number): void {
    if (amount <= this._currentHighestBid) {
      bidder.receiveNotification(`Bid rejected. Must be higher than $${this._currentHighestBid}`);
      return;
    }

    this._currentHighestBid = amount;
    this._highestBidder = bidder;

    for (const b of this._bidders.filter(b => b !== bidder)) {
      b.receiveNotification(`New highest bid: $${amount} by ${bidder.name}`);
    }

    console.log(`[AUCTIONEER] New highest bid: $${amount} by ${bidder.name}`);
  }

  announceWinner(): void {
    console.log('\n[AUCTIONEER] Auction ended!');

    if (!this._highestBidder) {
      console.log('[AUCTIONEER] No bids were placed. Item not sold.');
      return;
    }

    console.log(`[AUCTIONEER] Winner: ${this._highestBidder.name}`);
    console.log(`[AUCTIONEER] Winning bid: $${this._currentHighestBid}`);
    console.log(`[AUCTIONEER] Item: ${this._itemName}`);

    for (const bidder of this._bidders) {
      const message = bidder === this._highestBidder
        ? `Congratulations! You won ${this._itemName} for $${this._currentHighestBid}`
        : `${this._itemName} sold to ${this._highestBidder.name} for $${this._currentHighestBid}`;
      bidder.receiveNotification(message);
    }
  }
}
